src/NumericalAlgorithms/Interpolation/PredictedZeroCrossing.hpp

```diff
@@ -15,6 +15,30 @@ namespace intrp {
  *
  * Fits a linear function to a set of y_values at different x_values
  * and uses the fit to predict what x_value the y_value zero will be crossed.
+ * If we cannot determine (within the error bars of the fit) whether the
+ * zero crossing occurs for x_value < 0 versus x_value > 0, then we return
+ * zero.  Otherwise we return the best-fit x_value for when the function
+ * crosses zero.
+ *
+ * \details We fit to a straight line: y = intercept + slope*x.
+ * So our best guess is that the function will cross zero at
+ * x_best_fit = -intercept/slope.
+ *
+ * However, the data are assumed to be noisy.  The fit gives us error
+ * bars for the slope and the intercept.  Given the error bars, we can
+ * compute four limiting crossing values x0, x1, x2, and x3 by using
+ * the maximum and minimum possible values of slope and intercept.
+ * For example, if we assume slope<0 and intercept>0, then the
+ * earliest possible crossing consistent with the error bars is
+ * x3=(-intercept+delta_intercept)/(slope-delta_slope) and the latest
+ * possible crossing consistent with the error bars is
+ * x0=(-intercept-delta_intercept)/(slope+delta_slope).
+ *
+ * We compute all four crossing values and demand that all of them
+ * are either at x>0 (i.e. in the future if x is time) or at x<0
+ * (i.e. in the past if x is time).  Otherwise we conclude that we
+ * cannot determine even the sign of the crossing value, so we return
+ * zero.
```

Comment:
why this choice?  why not return the best_fit and the bounds and let the calling code decide what to do?  I'm confused as to why x=0 is special.  Are there constraints on the input values of x?

Reply:
Why x=0 is special: In actual usage, x=0 corresponds to the current time, and all the other input values of x will be negative (i.e. past times).  The y values are things that the control system needs to keep positive, like the min char speeds on the excision bdry, or the difference between the horizon and the excision boundary. So a crossing time with x>0 is in the future (and so the control system probably needs to take action), and a crossing time with x<0 is in the past (so the control system can be assured that it is safe).

And yes, if you want, this logic can be moved to ZeroCrossingPredictor (#4147) thus keeping PredictedZeroCrossing more general. Should I do that?

Reply:
I would keep PredictedZeroCrossing more general if it is not an implementation detail (in which case it should be put in a detail namespace).